src/viam/sdk/services/mlmodel/private/proto.cpp

```diff
@@ -0,0 +1,363 @@
+// Copyright 2023 Viam Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <viam/sdk/services/mlmodel/private/proto.hpp>
+
+#include <stack>
+
+#include <boost/variant/get.hpp>
+
+namespace viam {
+namespace sdk {
+namespace mlmodel_details {
+
+namespace {
+
+template <typename T>
+::grpc::Status pb_value_to_tensor_t(const MLModelService::tensor_info& tensor_info,
+                                    const ::google::protobuf::Value& pb,
+                                    tensor_storage* ts,
+                                    MLModelService::named_tensor_views* ntvs) {
+    // Stage our backing vector into `ts` and obtain a reference to it.
+    auto& storage_variant = *ts->emplace(ts->end(), std::vector<T>{});
+    auto& storage = boost::get<std::vector<T>>(storage_variant);
+
+    // Seed our shape vector as starting with one unknown dimension.
+    // We will populate it while walking the protobuf. Also create a
+    // stack of Value pointers which we will use to manage our
+    // walk. Start a depth counter so we know the index of the
+    // dimension we are iterating.
+    typename MLModelService::tensor_view<T>::shape_type shape{};
+    std::stack<const ::google::protobuf::Value*> vs{{&pb}};
+    size_t depth = 0;
+
+    while (!vs.empty()) {
+        if (!vs.top()) {
+            vs.pop();
+            --depth;
```

Comment:
The decrement at zero wouldn't be UB because unsigned integers wrap. So you would end up with a huge positive value. If you used that to index into `shape` that'd definitely be undefined behavior. In practice you would almost certainly observe a crash. That said, I'm not too worried about it getting out of sync because there is only one increment and one decrement, and they are paired fairly well with the marking of the top of the stack as nullptr. See also my comment above about the future of these loops.